Resilient deployment (RateLimiting)

Boomatang opened the pull request:
This cover resilient deployment for limitador.

closes:
- #1306 
- #1307 
- #1308 
- #1309 
- #1310 
- #1311 
- #1312 
- #1313 
- #1314 
- #1315 
- #1326 
- #1318 
- #1319 
- #1320 

For validation this pr implements the [Resilient Deployment of data plane components](https://docs.kuadrant.io/1.1.x/install-olm/#resilient-deployment-of-data-plane-components).

The main goal is to allow the user to have existing configuration, and not have kuadrant remove that configuration on upgrade. The user should be allowed to modify the configuration to suit their needs.

Boomatang committed:
ADD: Resilient Deployment Workflow Structure

Adds the skeleton structure of the workflow required for the resilient
deployment feature. Currently, there is only logging on the debug level.

The workflow is disabled if the kuadrant CR does not have the following
annotation.
`kuadrant.io/experimental-dont-use-resilient-data-plane: "true"`

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: isConfigured (counter storage)

Add the function to check it the counter storage has being configured.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: mirror counterStorage to limitador storage

When the `resilience.counterStorage` is configured in the kuadrant CR
the values are mirrored to the limitador CR `spec.storage`. This is a
one to one replacement.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: History

Set up a store that holds the past state.
Currently, the only resource held is the kuadrant resource.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: counterStorage UnSet

When the `counterStorage` is removed from the kuadrant resource it is also
removed from the limitador resource. If the user manually adds the
storage configuration to the limitador resource the change will not be
reverted if the feature is disabled in the kuadrant resource.

**EDGE CASE**
If the operator is scaled down when the user removes the
`counterStorage` from the kuadrant resource, and then restarts the
operator. The operator will not clean up the resources, it assumes it
was existing user configuration.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: integration test for counter storage

This adds integration tests for counter storage configuration
in the kuadrant resource. As there is interaction with the kuadrant
resource the test can **NOT** be run in parallel.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Resilient Deployment Counter Storage Documentation

Adds the initial documentation for the counter storage resilient
deployments.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
UPDATE: Kuadrant API

Add ratelimiting to the resiliencies object.

ADD: CEL for ratelimiting

This adds CEL validation to the configuring of the
resilience.rateLimiting. The user must state what method of storage they require in resilience.counterStorage.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Setting limitador replicas

Set the number of limitador replicas to 2.
2 is the initial default that we are picking.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Unset Replicas in limitador

Limitador does not revert the number of replicas back to one when the
replicas is unset in the limitador resource. So this sets the replicas
to one. This not a full reverting of the changes that was done.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Basic Status

Add a basic status for the resilience deployment. The status states if
the feature has be being configured or not.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Advanced Status

The status block of the kuadrant resource will state when the replicas
of the limitador resource has being modified when the
resilience.rateLimiting is enabled in the kuadrant CR.

There is a warning message when the number of replicas is below the
default, and a info message when the number is greater.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
REFACTOR: kuadrant_resource_test > counter_storage_test

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: integration test (resilience.RateLimiting)

Add a number of test to ensure the resilience.RateLimiting feature is
working as expected. This also cover the case where the user has an
existing configuration.

Test require updates in the kuadrant resource, and at time of writing
can not be run in parallel.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Pod Disruption Budget

Add  the ability to configure the pdb for limitador.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: limitador resource limits

Define resource limits that will be set in the limitador CR.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Set topologySpreadConstraints for limitador

This configures the spread constraints in the limitador deployment.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Unset limitador topology spread constraints

Undo the changes made to the limitador deployment.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: limitador topology spread constraints status

Add status block to the kuadrant resource when the limitador deployment
topology spread constraints don't match the default set by kuadrant.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
ADD: Integration tests

Add test around the topology spread constrains for limitador.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
FIX: issues found with integration tests

There were some issues found with running the integration tests.
Main problem was a timeout, so this has being bumped
A nil pointer was also found.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
FIX: Rebase issue

During the rebase the function name was changed or missed.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
FIX: issues found by PR Checks

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

maleck13 commented on api/v1beta1/kuadrant_types.go:
```diff
@@ -111,6 +127,24 @@ func (m *MTLS) IsAuthorinoEnabled() bool {
 	return m.Enable && ptr.Deref(m.Authorino, m.Enable)
 }
 
+// +kubebuilder:validation:XValidation:rule="has(self.rateLimiting) ? (self.rateLimiting == true && has(self.counterStorage)) || (self.rateLimiting == false && has(self.counterStorage)) || (self.rateLimiting == false && !has(self.counterStorage)) : true",message="resilience.counterStorage needs to be explictly configured when using resilience.rateLimiting."
```
```suggestion
// +kubebuilder:validation:XValidation:rule="has(self.rateLimiting) ? (self.rateLimiting == true && has(self.counterStorage)) || (self.rateLimiting == false && has(self.counterStorage)) || (self.rateLimiting == false && !has(self.counterStorage)) : true",message="resilience.counterStorage needs to be explicitly configured when using resilience.rateLimiting."
```

maleck13 commented on doc/user-guides/resilience/counterStorage.md:
```diff
@@ -0,0 +1,247 @@
+
+## Disabling Counter Storage
+
+When the counter storage configuration is removed, the Kuadrant Operator will revert the `Limirtador` CR back to the default of using In-Memory counters.
```
```suggestion
When the counter storage configuration is removed, the Kuadrant Operator will revert the `Limitador` CR back to the default of using In-Memory counters.
```

maleck13 commented on internal/controller/history.go:
```diff
@@ -0,0 +1,54 @@
+	kuadrantv1beta1 "github.com/kuadrant/kuadrant-operator/api/v1beta1"
+)
+
+type History struct {
```
Do we still need this?

Boomatang replied:
Reworked

maleck13 commented on internal/controller/kuadrant_status_updater.go:
```diff
@@ -145,6 +162,117 @@ func mtlsLimitador(kObj *kuadrantv1beta1.Kuadrant, state *sync.Map) *bool {
+	deployment := GetDeploymentForParent(topology, kuadrantv1beta1.LimitadorGroupKind)
+	if deployment != nil {
+		message := make([]string, 0)
+		for _, item := range deployment.Spec.Template.Spec.TopologySpreadConstraints {
```
Should we check if the length of topology constraints matches 2?

Boomatang replied:
Are the only allowed values for the topology spread constraints "kubernetes.io/hostname" and "kubernetes.io/zone"? If custom topology keys are allowed, we should not assume the list will only have a length of 2.

maleck13 commented on internal/controller/ratelimit_workflow_helpers.go:
```diff
@@ -54,7 +55,23 @@ func GetLimitadorFromTopology(topology *machinery.Topology) *limitadorv1alpha1.L
+	return limitador.DeepCopy()
+}
+
+func GetMachineryObjectFromTopology(topology *machinery.Topology, groupkind schema.GroupKind) machinery.Object {
```
This seems possibly in the wrong place? Seems a quite useful utility that might be used in many locations?

Boomatang replied:
moved

maleck13 commented on tests/common/kuadrant/counter_storage_test.go:
```diff
@@ -0,0 +1,193 @@
+			kuadrantKey := client.ObjectKey{Name: "kuadrant-sample", Namespace: testNamespace}
+			limitadorKey := client.ObjectKey{Name: kuadrant.LimitadorName, Namespace: testNamespace}
+
+			By("Apply kurdrant resouce with counterStorage configured")
```
```suggestion
			By("Apply kuadrant resource with counterStorage configured")
```

maleck13 commented on tests/common/kuadrant/rate_limiting_test.go:
```diff
@@ -0,0 +1,809 @@
+			Expect(err).NotTo(HaveOccurred())
+			Eventually(tests.KuadrantIsReady(testClient(), kuadrantKey)).WithContext(ctx).Should(Succeed())
+
+			By("limitador resource keeps initial configuration")
```
should we also check the status is correct on the kuadrant object?

eguzki commented on api/v1beta1/kuadrant_types.go:
```diff
@@ -71,6 +72,17 @@ func (k *Kuadrant) IsMTLSAuthorinoEnabled() bool {
+func (k *Kuadrant) BasicResilienceStatus() ResilienceStatus {
+	result := ResilienceStatus{}
+	result.RateLimiting = ptr.To((k.Spec.Resilience != nil && k.Spec.Resilience.RateLimiting))
+	if k.Spec.Resilience != nil && k.Spec.Resilience.CounterStorage != nil {
```
nit:
```go
result.CounterStorage = ptr.To(k.Spec.Resilience != nil && k.Spec.Resilience.CounterStorage != nil)
```

Boomatang replied:
reworked

maleck13 commented on tests/common/kuadrant/rate_limiting_test.go:
```diff
@@ -0,0 +1,809 @@
+				g.Expect(lObj.Spec.ResourceRequirements.Requests.Memory().Value()).To(Equal(userMemory.Value()))
+			},
+			).WithContext(ctx).Should(Succeed())
+			By("")
```
empty By() here. Were you planning on testing the kuadrant status here also?

maleck13 reviewed (commented):
Open questions on the history still. Also few minor comments. It is looking pretty good at this stage and once comments resolved and failing checks I will verify and we can look to merge

eguzki commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,564 @@
+		return false
+	}
+
+	if val, exists := k.GetAnnotations()[ResilienceFeatureAnnotation]; exists {
```
So, to enable this I have to define an annotation like this:

```yaml
annotations:
  kuadrant.io/experimental-dont-use-resilient-data-plane: true
```

Should the annotation name be changed?

Boomatang replied:
I am not sure if it needs to be changed. As soon as the all the work in the area is done the need for the annotation would be removed. This is most there so we can do iterative work in to main without affecting existing user. This is an undocumented feature.

eguzki commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,564 @@
+
+func NewResilienceDeploymentTasks(client *dynamic.DynamicClient) []controller.ReconcileFunc {
+	return []controller.ReconcileFunc{
+		NewResilienceAuthorizationReconciler().Subscription().Reconcile,
```
is the authorization resilient configuration in the scope of this PR?

eguzki commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,564 @@
+	}
+}
+
+func (r *ResilienceRateLimitingReconciler) reconcile(ctx context.Context, _ []controller.ResourceEvent, topology *machinery.Topology, _ error, state *sync.Map) error {
```
it would be nice to keep this method smaller. Too much done within it. This is not a request, just a "good to have" advice.

Boomatang replied:
I hear you. I guess this would be in the "Make it work" stage of "Make it work, Make it right, Make it right".

eguzki commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,564 @@
+
+	wasConfigured := false
+	if history.kuadrant != nil {
+		wasConfigured = history.kuadrant.Spec.Resilience.IsRateLimitingConfigured()
```
do we really need to know if in the past it was configured?

Usually controllers take actions based on the current state of the cluster. Having to code different paths from the previous steps to the current state might be unbearable and hard to maintain.

eguzki reviewed (commented):
left some comments

Boomatang commented on doc/user-guides/resilience/counterStorage.md:
```diff
@@ -0,0 +1,247 @@
+<!-- Reference: https://github.com/Kuadrant/limitador-operator/blob/main/doc/storage.md -->
+
+Rate limit counters are stored in a backend storage. 
+This is in contrast to the storage of the limits themselves, which are always stored in ephemeral memory. 
```
Far point, I didn't think there was need for any rewording as it is part of the limitador-operator docs. Does this mean that the state is also wrong in the original documentation.

Boomatang committed:
Remove: History Object

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
FIX: Liniter issues

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
FIX: integration test update

If there is an error writing the resource back to the cluster the
status for the resilience should not be updated to ensure that the next
iteration can correctly write back to the cluster.

FIX: logic on cleanup

The cleanup check would fail if the resource had failed to write to the
cluster on the first instances.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang commented on tests/common/kuadrant/rate_limiting_test.go:
```diff
@@ -0,0 +1,813 @@
+						RateLimiting:   true,
+						CounterStorage: &limitadorv1alpha1.Storage{},
+					},
+				}
+			})
+			Eventually(tests.KuadrantIsReady(testClient(), kuadrantKey)).WithContext(ctx).Should(Succeed())
+
+			By("Check the replica vaules in the limitador resource")
+			Eventually(func(g Gomega) {
+				limitadorKey := client.ObjectKey{Name: kuadrant.LimitadorName, Namespace: testNamespace}
+				lObj := &limitadorv1alpha1.Limitador{}
+				err := k8sClient.Get(ctx, limitadorKey, lObj)
+				g.Expect(err).ToNot(HaveOccurred())
+				g.Expect(*lObj.Spec.Replicas).To(Equal(controllers.LimitadorReplicas))
+			},
+			).WithContext(ctx).Should(Succeed())
+
+			By("Disable the rateLimiting feature")
+			Eventually(func(g Gomega) {
+				kObj := &kuadrantv1beta1.Kuadrant{}
+				err := k8sClient.Get(ctx, kuadrantKey, kObj)
+				g.Expect(err).ToNot(HaveOccurred())
+				kObj.Spec.Resilience.RateLimiting = false
+				err = k8sClient.Update(ctx, kObj)
+				g.Expect(err).NotTo(HaveOccurred())
+			}).WithContext(ctx).Should(Succeed())
+			Eventually(tests.KuadrantIsReady(testClient(), kuadrantKey)).WithContext(ctx).Should(Succeed())
+
+			By("Check the replica vaules in the limitador resource have being reverted")
+			Eventually(func(g Gomega) {
+				limitadorKey := client.ObjectKey{Name: kuadrant.LimitadorName, Namespace: testNamespace}
+				lObj := &limitadorv1alpha1.Limitador{}
+				err := k8sClient.Get(ctx, limitadorKey, lObj)
+				g.Expect(err).ToNot(HaveOccurred())
+				g.Expect(*lObj.Spec.Replicas).To(Equal(1))
+			},
+			).WithContext(ctx).Should(Succeed())
+
+		}, testTimeOut)
+	})
```
Root cause of this issue was found, there was a logic bug in the recenciler checks.

Boomatang commented on internal/controller/state_of_the_world.go:
```diff
@@ -649,6 +650,7 @@ func initWorkflow(client *dynamic.DynamicClient) *controller.Workflow {
 		Precondition: NewEventLogger().Log,
 		Tasks: []controller.ReconcileFunc{
 			NewTopologyReconciler(client, operatorNamespace).Reconcile,
+			InitializeHistoryFunc,
```
Reworked

maleck13 commented on internal/controller/kuadrant_status_updater.go:
```diff
@@ -145,6 +244,117 @@ func mtlsLimitador(kObj *kuadrantv1beta1.Kuadrant, state *sync.Map) *bool {
+		return nil, remove
+	}
+
+	if lObj.Spec.Replicas != nil && *lObj.Spec.Replicas < LimitadorReplicas {
```
have you considered adding some functions here such as something like `userModifiedReplicas(dep *Deployment, expected int)(bool, string)` where the bool represents if it user changed and the string the message to use?

maleck13 commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,625 @@
+	logger.V(level).Info("ResilienceDeployment Precondition", "status", "started")
+	defer logger.V(level).Info("ResilienceDeployment Precondition", "status", "completed")
+
+	state.Store(ExperimentalResilienceFeature, isExperimentalFeatureEnabled(topology))
```
would it be better to register an errorHandler here and just return a known error if the precondition fails?

Boomatang replied:
Possibly. My mental issue with doing that is saying if the experimental feature is disabled, that is an error. The question that runs around my head is how long do you expect this feature to be experimental? Is configuring a errorHandler the best use of our time?

maleck13 commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,625 @@
+	}
+}
+
+func (r *ResilienceAuthorizationReconciler) reconcile(ctx context.Context, _ []controller.ResourceEvent, _ *machinery.Topology, _ error, state *sync.Map) error {
```
one thought I had here was the resourceEvents contain the old and new object so perhaps these could be used to compare what has changed. Perhaps not needed 🤷

Boomatang replied:
I think it will give you the old and new version of the object. This however can not be used as the old version that you will be give is the old version on the cluster, and the last version you successfully reconciled. If you could compare the new version with the past reconciled version, then yes we could have used that.

maleck13 replied:
Ok so what exactly does it give you then?

maleck13 commented on internal/controller/kuadrant_status_updater.go:
```diff
@@ -111,19 +117,112 @@ func (r *KuadrantStatusUpdater) updateKuadrantStatus(ctx context.Context, kObj *
+	}
+
+	requests := false
+	if lObj.Spec.ResourceRequirements != nil && lObj.Spec.ResourceRequirements.Requests != nil {
```
I am a bit confused by this check. Do we not also need to check the kuadrant object? Seems these would be set if the kuadrant resilience had been defined? Seems like we are saying it user defined if set at all on the limitador obj?

maleck13 replied:
So checking this locally. Changing the values manually. I never get to see `rateLimiting: user defined`

```yaml
conditions:
  - lastTransitionTime: "2025-06-16T09:45:41Z"
    message: Deployment has unavailable replicas
    reason: LimitadorNotReady
    status: "False"
    type: Ready
  - lastTransitionTime: "2025-06-16T09:40:56Z"
    message: Number of Limitador replicas (3) greater than minimum default
    reason: UserModifiedLimitadorReplicas
    status: Unknown
    type: Resilience.Info.RateLimiting.Replicas
  - lastTransitionTime: "2025-06-16T09:44:04Z"
    message: Limitador recource Pod Disruption Budget differs from default configuration
    reason: UserModifiedLimitadorPodDisruptionBudget
    status: Unknown
    type: Resilience.Info.RateLimiting.PBD
  - lastTransitionTime: "2025-06-16T09:45:41Z"
    message: Limitador depoloyment TopologySpreadConstraints for key "kubernetes.io/zone"
      is user modified.
    reason: UserModifiedLimitadorTopologySpreadConstraints
    status: Unknown
    type: Resilience.Info.RateLimiting.TopologySpreadConstraints
  mtlsAuthorino: false
  mtlsLimitador: false
  observedGeneration: 10
  resilience:
    counterStorage: kuadrant defined
    rateLimiting: kuadrant defined  # remains as kuadrant defined??
```

maleck13 commented on internal/controller/kuadrant_status_updater.go:
```diff
@@ -111,19 +117,112 @@ func (r *KuadrantStatusUpdater) updateKuadrantStatus(ctx context.Context, kObj *
+
+	resilienceError := false
+	stateResilienceError, ok := state.Load("resilienceError")
+	if ok {
+		resilienceError = stateResilienceError.(bool)
+	}
+
+	if resilienceError {
+		return kObj.Status.Resilience
+	}
+
```
This block reads oddly. If there is a resilience error we just return whatever is in the existing kuadrant status and don't report it? I feel like I must be making a mistake here? Also this piece of code could be simplified right?

```
if ok && stateResilienceError.(bool){
// do something with status
  return kObj.Status.Resilience
}
```

maleck13 replied:
Yes so testing this if an error occurs the status is not changed at all. Which means it is not clear to the end user

````
pec:
  resilience:
    counterStorage:
      redis:
        configSecretRef:
          name: redis-credentials
    rateLimiting: true
status:
  conditions:
  - lastTransitionTime: "2025-06-13T10:21:25Z"
    message: Kuadrant is ready
    reason: Ready
    status: "True"
    type: Ready
  mtlsAuthorino: false
  mtlsLimitador: false
  observedGeneration: 1
  resilience:
    counterStorage: kuadrant defined
    rateLimiting: kuadrant defined
```

I think we should be changing the ready condition here. Kuadrant is not ready in the way the user has asked for it to be configured 

```
  - lastTransitionTime: "2025-06-13T10:21:25Z"
    message: Failed to configure resilience for ratelimitng 
    reason: NotReady
    status: "False"
    type: Ready

Boomatang replied:
The reading of the block being odd is fair, will address. It is correct the condition type `Ready` should be updated while the reconcile is looping. Will look to address.

Boomatang commented on api/v1beta1/kuadrant_types.go:
```diff
@@ -153,6 +188,12 @@ func (r *KuadrantStatus) Equals(other *KuadrantStatus, logger logr.Logger) bool
 
+	if !reflect.DeepEqual(r.Resilience, other.Resilience) {
+		diff := cmp.Diff(r.Resilience, other.Resilience)
+		logger.V(0).Info("Resilience not equal", "difference", diff)
```
It should have being the same, missed it in my review.

Boomatang commented on internal/controller/kuadrant_status_updater.go:
```diff
@@ -145,6 +244,117 @@ func mtlsLimitador(kObj *kuadrantv1beta1.Kuadrant, state *sync.Map) *bool {
+		remove = append(remove, ResilienceInfoRRConditionType)
+	}
+
+	if lObj.Spec.PodDisruptionBudget != nil && lObj.Spec.PodDisruptionBudget.MaxUnavailable != nil && lObj.Spec.PodDisruptionBudget.MaxUnavailable.IntVal != LimitadorPDB {
```
This one does make more sense than the replicas above because of the `else if`. I will look at a bit of a tidy on it.  The issue with the object in question is the limitador CR and not a deployment CR still applies.

Boomatang committed:
UPDATE: PR comments

A number of updates base on the PR comments.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

maleck13 commented on internal/controller/kuadrant_status_updater.go:
```diff
@@ -23,7 +25,11 @@ import (
 const (
-	ReadyConditionType string = "Ready"
+	ReadyConditionType               string = "Ready"
+	ResilienceInfoRRConditionType    string = "Resilience.Info.RateLimiting.Replicas"
```
The value

maleck13 commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,627 @@
+			lObj.Spec.ResourceRequirements = &corev1.ResourceRequirements{}
+		}
+
+		if lObj.Spec.ResourceRequirements.Requests.Cpu().Value() == 0 {
+			if lObj.Spec.ResourceRequirements.Requests == nil {
```
This doesn't look right. We do a nil check after using the object in the previous check?

maleck13 commented:
**Verification**

I did the following things post code review.

✅  Create Kuadrant CR with no resilience ensure deploys and becomes ready. I did wonder if it may be better to just leave resilience status empty in this situation.

```yaml
status:
  conditions:
  - lastTransitionTime: "2025-06-20T10:57:15Z"
    message: Kuadrant is ready
    reason: Ready
    status: "True"
    type: Ready
  mtlsAuthorino: false
  mtlsLimitador: false
  observedGeneration: 1
  resilience:
    counterStorage: undefined
    rateLimiting: undefined

✅  Update the replicas and resources on limitador directly and see that the kuadrant resource reports correctly

```yaml
status:
  conditions:
  - lastTransitionTime: "2025-06-20T11:09:33Z"
    message: Deployment does not have minimum availability.
    reason: LimitadorNotReady
    status: "False"
    type: Ready
  mtlsAuthorino: false
  mtlsLimitador: false
  observedGeneration: 1
  resilience:
    counterStorage: undefined
    rateLimiting: user defined

✅ Define resilience true and counter storage via kuadrant see that limitador is updated to our defaults

```yaml
 spec:
    resilience:
      counterStorage:
        redis:
          configSecretRef:
            name: redis-credentials
      rateLimiting: true
  status:
    conditions:
    - lastTransitionTime: "2025-06-20T11:14:49Z"
      message: Kuadrant is ready
      reason: Ready
      status: "True"
      type: Ready
    mtlsAuthorino: false
    mtlsLimitador: false
    observedGeneration: 2
    resilience:
      counterStorage: kuadrant defined
      rateLimiting: kuadrant defined
---
spec:
    pdb:
      maxUnavailable: 1
    replicas: 2
    resourceRequirements:
      requests:
        cpu: 10m
        memory: 10Mi
    storage:
      redis:
        configSecretRef:
          name: redis-credentials
```
❌  change limitador directly and see we don't overwrite it again and we report status correctly

I expected a status warning here that the resources were not as expected

```yaml
spec:
    pdb:
      maxUnavailable: 1
    replicas: 2
    resourceRequirements:
      requests:
        cpu: 5m
        memory: 10Mi

---
status:
    conditions:
    - lastTransitionTime: "2025-06-20T11:30:52Z"
      message: Kuadrant is ready
      reason: Ready
      status: "True"
      type: Ready
    mtlsAuthorino: false
    mtlsLimitador: false
    observedGeneration: 2
    resilience:
      counterStorage: kuadrant defined
      rateLimiting: kuadrant defined
```

Boomatang committed:
REFACTOR: nil

Change the logic around the nil check

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

maleck13 commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,623 @@
+		if lObj.Spec.ResourceRequirements.Requests.Cpu().Value() == 0 {
+			cpu, err := resource.ParseQuantity(Resource10m)
+			if err != nil {
+				logger.Error(err, "failed to parse resurce cpu string", "status", "error")
```
should we return the error here rather than attempt to update ? I see it is highly unlikely given we are parsing a constant but it seems odd to just log and attempt to update anyway?

maleck13 commented on internal/controller/resilient_deployment_workflow.go:
```diff
@@ -0,0 +1,623 @@
+
+		if lObj.Spec.ResourceRequirements.Requests.Memory().Value() == 0 {
+			memory, err := resource.ParseQuantity(Resource10Mi)
+			if err != nil {
```
same as above

Boomatang committed:
FIX: Clean up

When the whole resilience block was removed the clean up of resource did
not trigger correctly.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
FIX: status going to user defined

If the number of replicas is removed from the limitador CR, the
limitador will not scale the number of replicas to the default of one.

We need to update the limitador CR twice, and in order to do this a
sleep was added.

There could be a better way to do this over adding a sleep, but I
couldn't find a reliable way.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang committed:
UPDATE: Kuadant CR API

Add the resilience.counterStorage to the kuadrant spec.

Signed-off-by: Jim Fitzpatrick <jfitzpat@redhat.com>

Boomatang commented:
Wont be done
